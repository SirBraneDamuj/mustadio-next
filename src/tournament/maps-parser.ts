const mapRegex = /^(?<mapNumber>\d+)\) (?<mapTitle>.*)$/;

type Map = {
  number: string;
  title: string;
};

export function parseMapsData(data: string) {
  let delimiter = "\r\n";
  if (data.indexOf(delimiter) == -1) {
    delimiter = "\n";
  }
  return data.split(delimiter).reduce((maps, mapLine) => {
    const regexResult = mapRegex.exec(mapLine);
    if (regexResult) {
      if (!regexResult.groups) {
        throw new Error("Maps Regex did not match expected groups");
      }
      const { mapNumber, mapTitle } = regexResult.groups;
      maps.push({
        number: mapNumber,
        title: mapTitle,
      });
    }
    return maps;
  }, [] as Map[]);
}
