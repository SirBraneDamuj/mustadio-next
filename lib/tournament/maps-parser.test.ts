import { promises as fs } from "fs";
import path from "path";
import { describe, expect, it } from "vitest";
import { parseMapsData } from "./maps-parser";

describe("parseMapsData", () => {
  it("parses maps.txt correctly (snapshot)", async () => {
    const filePath = path.join(
      __dirname,
      "../../resources/fftbg_fake/fake_tournament/maps.txt"
    );
    const file = await fs.readFile(filePath, "utf-8");
    const result = parseMapsData(file);

    expect(result).toMatchSnapshot();
  });
});
