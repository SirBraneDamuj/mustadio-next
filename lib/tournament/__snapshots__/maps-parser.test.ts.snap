// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseMapsData > parses maps.txt correctly (snapshot) 1`] = `
[
  {
    "number": "98",
    "title": "Gate of Lesalia Castle",
  },
  {
    "number": "115",
    "title": "Banished Fort",
  },
  {
    "number": "58",
    "title": "Underground Book Storage Second Floor",
  },
  {
    "number": "102",
    "title": "Tutorial Bridge",
  },
  {
    "number": "4",
    "title": "Lesalia Castle",
  },
  {
    "number": "36",
    "title": "Church Outside of Town",
  },
  {
    "number": "86",
    "title": "Doguola Pass",
  },
  {
    "number": "47",
    "title": "Zarghidas Trade City",
  },
]
`;
